### PR TITLE
MM-41185,MM-41065: fixes mark thread as unread

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

Until now we mark a thread as read/unread by sending a timestamp to the
server. This created some issues when marking a thread as unread from a
post. We needed to send the post.create_at - 1 as a timestamp but we
didn't do so consistently.

A new API to set a thread as unread is added to the server so that you
can mark a thread as unread by post id. Making all clients agnostic of
the timestamp, and thus solving consistency issues.

This commit adds the new API to the webapp client. And uses it to mark a
thread as unread from the post's dot menu.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-41185
https://mattermost.atlassian.net/browse/MM-41065
https://mattermost.atlassian.net/browse/MM-42473

#### Related Pull Requests

- Has server changes (https://github.com/mattermost/mattermost-server/pull/19951)

#### Release Note

```release-note
* Fixes issue where marking last post in a thread as unread didn't mark the thread as unread.
```

### CODE REVIEW

**Review by ashishbhate — APPROVED** [review]

**Comment by koox00** [comment]
@ashishbhate, @lieut-data I just wonder if instead fixing this in the webapp, we have to make the fix on the server.
To explain myself:
Scenario a thread containing a mention in the last reply. 
currently, without this fix, when marking a thread as last viewed at the last post's `create_at` the server would return `unread_mentions` to 1 BUT `unread_replies` to 0.

**Comment by ashishbhate** [comment]
I took a quick look at the thread store.

For various unreads and counts check, we use: 
`sq.Expr("ThreadMemberships.LastViewed < Threads.LastReplyAt")`,
`sq.Expr("Posts.CreateAt > ThreadMemberships.LastViewed")`, or 
`sq.Gt{"Posts.CreateAt": threadMembership.LastViewed}` 


While for the mention count checking we fetch posts with the condition:
```
Select("*").
From("Posts").
Where(sq.Eq{"RootId": threadId}).
Where(sq.Eq{"DeleteAt": 0}).
Where(sq.GtOrEq{"UpdateAt": since}).ToSql()
```

Notice the `GtOrEq` check here, vs the `Gt` checks for the unreads case. I think this might be causing the issue you're seeing.

**Comment by lieut-data** [comment]
> @ashishbhate, @lieut-data I just wonder if instead fixing this in the webapp, we have to make the fix on the server.

@koox00, I do like the idea of handling this on the server vs. knowing to pass in `create_at - 1`. I wonder if we might instead expose and use an API that amounts to `markThreadAsUnreadByPostId` and have the server take care of "all" the magic?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Blocking review while we discuss the approach. (Feel free to re-request as needed!)

**Comment by ashishbhate** [comment]
I'm in favour of a fix on the server side as well. Having to pass `post.create_at - 1` is an implementation detail that the webapp shouldn't need to know. The webapp should just be able to say, "mark thread as unread from this post"

**Comment by koox00** [comment]
> @koox00, I do like the idea of handling this on the server vs. knowing to pass in create_at - 1. I wonder if we might instead expose and use an API that amounts to `markThreadAsUnreadByPostId` and have the server take care of "all" the magic?

@lieut-data, @ashishbhate sorry for taking so long to reply, I got absorbed in mobile stuff.
I'm all in favor of using an API like `markThreadAsUnreadByPostId`, which is similar to what we do in channels btw.
I'll create a new PR to address that in the server.

**Comment by koox00** [comment]
I'll keep this one to make the necessary changes to conform to the new API when ready

**Review by lieut-data — APPROVED** [review]
:tada:

File: actions/post_actions.jsx [hunk 1]
```diff
@@ -259,7 +259,7 @@ export function markPostAsUnread(post, location) {
             const threadId = post.root_id || post.id;
             ThreadActions.handleFollowChanged(dispatch, threadId, currentTeamId, true);
             dispatch(manuallyMarkThreadAsUnread(threadId, post.create_at));
```
[the remark this answers is not in the document]
Comment (koox00) [reply actions/post_actions.jsx]: yep, thx! changed it :)

File: actions/post_actions.jsx [hunk 2]
```diff
@@ -258,7 +258,7 @@ export function markPostAsUnread(post, location) {
         if (isCollapsedThreadsEnabled(state) && (location === 'RHS_ROOT' || location === 'RHS_COMMENT')) {
             const threadId = post.root_id || post.id;
             ThreadActions.handleFollowChanged(dispatch, threadId, currentTeamId, true);
-            dispatch(manuallyMarkThreadAsUnread(threadId, post.create_at));
+            dispatch(manuallyMarkThreadAsUnread(threadId, post.create_at - 1));
```
Comment (lieut-data) [inline actions/post_actions.jsx:261]: Should we... add an explicit API for this? ;P

**Review by lieut-data — APPROVED** [review]

**Pull request merged**